message_ix/tests/test_feature_capacity_factor.py

```diff
@@ -7,29 +7,39 @@
 
 def check_solution(scen: Scenario) -> None:
     """Perform several assertions about the solution of `scen`."""
-    # Reading "ACT" and "CAP" from the solution
+    # Reading "ACT" and "CAP" from the solution, plus "operation_factor"
     act = scen.var("ACT").set_index(["technology", "time"])
     cap = scen.var("CAP").set_index(["technology"])
+    op_fac = scen.par("operation_factor").set_index(["technology"])
 
-    # 1) ACT is zero when capacity factor is zero
+    # 1) ACT is zero when capacity factor (CF) is zero
     cf = scen.par("capacity_factor").set_index(["technology", "time"])
     cf_zero = cf.loc[cf["value"] == 0]
     for i in cf_zero.index:
         assert act.loc[i, "lvl"] == 0
 
-    # 2) CAP is correctly calculated based on ACT and capacity_factor
-    for i in act.loc[act["lvl"] > 0].index:
-        # Correct ACT based on duration of each time slice
+    # 2) CAP is correctly calculated based on "ACT", CF, and "operation_factor"
+    for i in act.index:
+        # Correct CF based on duration of each time slice
         duration = float(scen.par("duration_time", {"time": i[1]})["value"])
-        act.loc[i, "duration-corrected"] = act.loc[i, "lvl"] / duration
-        # Divide by (non-zero) capacity factor
-        act.loc[i, "cf-corrected"] = act.loc[i, "duration-corrected"] / float(
-            cf.loc[i, "value"]
-        )
+        if i[1] != "year":
+            cf.loc[i, "duration-corrected"] = cf.loc[i, "value"] * duration
+            act.loc[i, "cf-corrected"] = act.loc[i, "lvl"] / float(
```

Comment:
This seems to cause the `RuntimeWarning`. I'm wondering about `i` and the hardcoded index `1`: is that always correct or does it assume a certain structure of `i`? If so, can we change to using no assumption?

Reply:
The problem arises in the `test_capacity_factor_zero_two`, where the `solar_pv_ppl` has a `capacity_factor` of `0` in the time slice `night`. In
```
if i[1] != "year":
            cf.loc[i, "duration-corrected"] = cf.loc[i, "value"] * duration
            act.loc[i, "cf-corrected"] = act.loc[i, "lvl"] / float(
                cf.loc[i, "duration-corrected"]
            )
```
this gives us a division by 0 since `cf.loc[i, "value"]` is 0. So we have to provide some condition/prescription on how to deal with cases like this.

Reply:
I've now added a condition that checks if either `duration` or `cf.loc[i, "value"]` are `0`. In that case, we now set `act.loc[i, "cf-corrected"]` to `act.loc[i, "lvl"]`, so we don't change it. This relies on the assumption that in this case, `act.loc[i, "lvl"]` will also be 0, which is implied (and asserted) by the test in line 19. 
All the tests are still passing for me, but there's no `RuntimeWarning` anymore.

Reply:
Thanks @glatterf42 for reviewing this and suggesting improvements. Indeed, as you noted, zero capacity factors are tested and they can be excluded from the follow-up calculations. The index `i` in that loop is a tuple with two members, the first one being the technology name and the second time slices. So, the index `1` is always related to time.